### ISSUE TITLE
accelerate.cpu_offload can not be used for training

### ISSUE THREAD

**Issue by CO18326** [huggingface/accelerate#3785]
### System Info

```Shell
- `Accelerate` version: 1.10.0
- Platform: Linux-6.14.0-27-generic-x86_64-with-glibc2.39
- `accelerate` bash location: /home/ub-11/miniconda3/envs/torch/bin/accelerate
- Python version: 3.13.5
- Numpy version: 2.3.2
- PyTorch version: 2.8.0+cu128
- PyTorch accelerator: CUDA
- System RAM: 125.57 GB
- GPU type: NVIDIA RTX A4000
- `Accelerate` default config:
        Not found
```

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [x] My own task or dataset (give details below)

### Reproduction

import time
import argparse
import torch
from torch.utils.data import DataLoader
from datasets import load_dataset
from transformers import (
    AutoTokenizer,
    AutoModelForCausalLM,
    default_data_collator,
    )
from accelerate.utils import named_module_tensors
  

from torch.optim import AdamW
from accelerate import Accelerator, cpu_offload
import torch.cuda.nvtx as nvtx
import psutil
from accelerate.state import AcceleratorState
def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--seq_len", type=int, default=43)
    parser.add_argument("--steps", type=int, default=3)
    args = parser.parse_args()

    accelerator = Accelerator()
    device = torch.device("cuda")

    model_name = "gpt2-large"
    print("Loading tokenizer and model...")
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    tokenizer.pad_token = tokenizer.eos_token

    model = AutoModelForCausalLM.from_pretrained(model_name)
    model = model.to(device)
    optimizer = AdamW(model.parameters(), lr=5e-5)

   
    model = cpu_offload(model, device)

   
    dataset = load_dataset("wikitext", "wikitext-2-v1", split="train")

    def tokenize_fn(examples):
        model_inputs = tokenizer(
            examples["text"],
            truncation=True,
            padding="max_length",
            max_length=args.seq_len,
        )
        labels = []
        for seq in model_inputs["input_ids"]:
            labels.append([tok if tok != tokenizer.pad_token_id else -100 for tok in seq])
        model_inputs["labels"] = labels
        return model_inputs

    dataset = dataset.map(tokenize_fn, batched=True, remove_columns=["text"])

    dataloader = DataLoader(
        dataset,
        batch_size=2,
        shuffle=True,
        collate_fn=default_data_collator,
    )
    
    optimizer, dataloader = accelerator.prepare(optimizer, dataloader)
    state = AcceleratorState()

    for i, group in enumerate(optimizer.param_groups):
        for j, p in enumerate(group["params"]):
            print(f"group {i} param {j}: id={id(p)}, addr={p.data.data_ptr()}, grad={p.grad}, data[0,0]={p.data.view(-1)[0].item():.6f}")
            break

     

    model.train()
    step_times = []

   
        
    for step, batch in enumerate(dataloader):
        start = time.time()
        outputs = model(**batch)
        loss = outputs.loss
                        
        loss.backward()

        for i, group in enumerate(optimizer.param_groups):
            for j, p in enumerate(group["params"]):
                print(f"group {i} param {j}: id={id(p)}, addr={p.data.data_ptr()}, grad={p.grad}, data[0,0]={p.data.view(-1)[0].item():.6f}")
                break
        optimizer.step()

        for i, group in enumerate(optimizer.param_groups):
            for j, p in enumerate(group["params"]):
                print(f"group {i} param {j}: id={id(p)}, addr={p.data.data_ptr()}, grad={p.grad}, data[0,0]={p.data.view(-1)[0].item():.6f}")
                break
                        
        optimizer.zero_grad()
        step_time = time.time() - start

        mem = torch.cuda.memory_allocated() / (1024 ** 3)
        cpu_mem = psutil.Process().memory_info().rss / 1024**3
        print(f"Step {step} | Loss={loss.item():.4f} | GPU={mem:.2f} GB | CPU={cpu_mem:.2f} GB | {step_time:.2f}s")

        if step + 1 >= args.steps:
            break
                
            

            

if __name__ == "__main__":
   main()
    

### Expected behavior

During our research while  experimenting with accelerate.cpu_offload  we figured out that it can not be used for training although the training script runs without any error.

1. The mechanism the accelerate.cpu_offload  uses is the reason for it, as it maintains the weight_map dict in the cpu side which stores the original parameter of model
2. During forward pass it loads those values to execution device.
3. Hence computation graph hold the refrences to the current tensors that are on GPU
4. After forward pass model parameters are moved to meta.
5. optimiser.step tries to updates the model.parameters but it holds the references to old model parameters objects when model was initlised, and not see any gradient writen on them, hence no update.

-> conclusion : forward pass is happening using different parameters, gradient is written on different parameters, and optimiser.step is happening on different parameter.
->Interesting point is script dont give any error.

  `  
   
    model = AutoModelForCausalLM.from_pretrained(model_name)
    optimizer = AdamW(model.parameters(), lr=5e-5)
    model = cpu_offload(model, torch.device("cuda"))

    dataset = load_dataset("wikitext", "wikitext-2-v1", split="train")

    def tokenize_fn(examples):
        model_inputs = tokenizer(
            examples["text"],
            truncation=True,
            padding="max_length",
            max_length=args.seq_len,
        )
        labels = []
        for seq in model_inputs["input_ids"]:
            labels.append([tok if tok != tokenizer.pad_token_id else -100 for tok in seq])
        model_inputs["labels"] = labels
        return model_inputs

    dataset = dataset.map(tokenize_fn, batched=True, remove_columns=["text"])
    dataloader = DataLoader(dataset, batch_size=2, shuffle=True, collate_fn=default_data_collator)
    optimizer, dataloader = accelerator.prepare(optimizer, dataloader)

    model.train()
    for step, batch in enumerate(dataloader):
        outputs = model(**batch)
        loss = outputs.loss
        loss.backward()
        optimizer.step()
        optimizer.zero_grad()
        if step + 1 >= args.steps:
            break`

By logging the parameter objects referenced by optimiser , it can be validated that gradients are none , and parameter values are not updating:
<img width="1221" height="366" alt="Image" src="https://github.com/user-attachments/assets/1e0cff9e-caa5-46f2-90a5-efb6501e732c" />

The following script demonstrates what is essentially happening 

[test_script](https://github.com/user-attachments/files/22250322/test-test.py)

`
